File: php/src/Catalog.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * The read projection — ported from `Catalog` in lib/golden_record_core.ex.
 *
 * `project` folds the per-lane members + the current claim view into customer-facing records:
 * products ▸ variants ▸ {codes, survivorship attributes, product label, media, categories,
 * substances, derived descriptions}. Visibility is DERIVED at read time — edges are resolved to
 * their current owner key on every read, so a merge/split converges every product's view with no
 * writes. On the 422156 path: media arrives via `depicts` edges (MED_* records), descriptions via
 * `describes` edges (DSC_* records), categories via `member_of` edges, substances are empty.
 */
final class Catalog
{
    /**
     * @param array<string, array<string, array{0: string, 1: string}>> $members key => code-set (all lanes)
     * @param list<Claim> $liveClaims current claim view
     * @param array{attr: array<string, mixed>, product: array<string, mixed>} $overrides
     * @return list<GoldenRecord>
     */
    public static function project(array $members, array $liveClaims, Priority|callable $priority, array $overrides): array
    {
        $lanes = Lanes::partitionMembers($members);
        $attrs = self::filterKind($liveClaims, 'attribute');
        $groups = self::filterKind($liveClaims, 'grouping');
        $media = self::filterKind($liveClaims, 'media');
        $edges = self::filterKind($liveClaims, 'edge');

        // Flatten each lane's members once — owner() scanned every member set per edge endpoint.
        $mediaIndex = self::ownerIndex($lanes['media']);
        $subIndex = self::ownerIndex($lanes['substance']);
        $descIndex = self::ownerIndex($lanes['description']);

        $variants = [];
        foreach ($lanes['product'] as $key => $codes) {
            $substances = self::resolveSubstances($codes, $edges, $subIndex);
            $variants[] = new Variant(
                $key,
                Sets::valuesSorted($codes),
                self::resolveAttributes($key, $codes, $attrs, $priority, $overrides['attr']),
                self::resolveProduct($key, $codes, $groups, $priority, $overrides['product']),
                array_merge(
                    self::resolveMedia($codes, $media, $priority),
                    self::resolveDepicted($codes, $edges, $lanes['media'], $mediaIndex, $attrs, $priority)
                ),
                self::resolveCategories($codes, $edges),
                $substances,
                self::resolveDescriptions($codes, $edges, $lanes, $subIndex, $descIndex, $substances, $attrs, $priority),
            );
        }

        // Group variants by their product label, sort products by label, variants by key.
        /** @var array<string, array{product: mixed, variants: list<Variant>}> $byProduct */
        $byProduct = [];
        $order = [];
        foreach ($variants as $v) {
            $pk = self::scalarKey($v->product->value);
            if (!isset($byProduct[$pk])) {
                $byProduct[$pk] = ['product' => $v->product->value, 'variants' => []];
                $order[] = $pk;
            }
            $byProduct[$pk]['variants'][] = $v;
        }

        $products = [];
        foreach ($order as $pk) {
            $products[] = $byProduct[$pk];
        }
        usort($products, static fn (array $a, array $b): int => self::compareProductValues($a['product'], $b['product']));

        $records = [];
        foreach ($products as $p) {
            $ordered = $p['variants'];
            usort($ordered, static fn (Variant $a, Variant $b): int => strcmp($a->key, $b->key));
            $records[] = new GoldenRecord($p['product'], $ordered);
        }

        return $records;
    }

    /**
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $attrs
     * @param array<string, mixed> $attrOverrides
     * @return list<array{0: string, 1: Decision}>
     */
    private static function resolveAttributes(string $key, array $codes, array $attrs, Priority|callable $priority, array $attrOverrides): array
    {
        // $key/$attrOverrides are unused pending steward-override support (Elixir's apply_override
        // was deliberately dropped on the 422156 path) — kept for the Elixir-parity signature.
        return self::laneAttributes($codes, $attrs, $priority);
    }

    /**
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $groups
     * @param array<string, mixed> $productOverrides
     */
    private static function resolveProduct(string $key, array $codes, array $groups, Priority|callable $priority, array $productOverrides): Decision
    {
        if (array_key_exists($key, $productOverrides)) {
            return new Decision($productOverrides[$key], 'steward', 'resolved_by_steward', []);
        }

        return self::resolveProductFromClaims($codes, $groups, $priority);
    }

    /**
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $groups
     */
    private static function resolveProductFromClaims(array $codes, array $groups, Priority|callable $priority): Decision
    {
        $entries = [];
        foreach ($groups as $g) {
            if (Sets::member($codes, $g->data['code'])) {
                $entries[] = ['source' => $g->source, 'value' => $g->data['product'], 'order' => $g->order];
            }
        }

        if ($entries === []) {
            return new Decision(['none', '—'], null, 'resolved', []);
        }

        $base = Survivorship::decide('product', $entries, $priority);
        $distinct = [];
        foreach ($entries as $e) {
            $distinct[self::scalarKey($e['value'])] = true;
        }
        if (count($distinct) > 1) {
            return new Decision(null, null, 'needs_review', $base->candidates);
        }

        return $base;
    }

    /**
     * Legacy media-claim resolution (dedup by asset, highest-priority source wins).
     *
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $media
     * @return list<array<string,mixed>>
     */
    private static function resolveMedia(array $codes, array $media, Priority|callable $priority): array
    {
        /** @var array<string, list<Claim>> $byAsset */
        $byAsset = [];
        foreach ($media as $m) {
            if (Sets::member($codes, $m->data['target'])) {
                $byAsset[self::scalarKey($m->data['asset'])][] = $m;
            }
        }

        $rank = Survivorship::rankFn($priority);

        $out = [];
        foreach ($byAsset as $claims) {
            $best = $claims[0];
            $bestRank = $rank('media', $best->source);
            foreach ($claims as $m) {
                $r = $rank('media', $m->source);
                if ($r < $bestRank) {
                    $bestRank = $r;
                    $best = $m;
                }
            }
            $out[] = [
                'asset' => $best->data['asset'],
                'role' => $best->data['role'],
                'source' => $best->source,
                'uri' => $best->data['uri'],
            ];
        }

        usort($out, static function (array $a, array $b): int {
            return [$a['role'] !== 'primary', self::scalarKey($a['asset'])]
                <=> [$b['role'] !== 'primary', self::scalarKey($b['asset'])];
        });

        return $out;
    }

    /**
     * Categories via `member_of` edges: the collection {namespace, member} pairs this record's
     * codes point at, unique + sorted.
     *
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $edges
     * @return list<array{0: string, 1: string}>
     */
    private static function resolveCategories(array $codes, array $edges): array
    {
        $seen = [];
        $out = [];
        foreach ($edges as $e) {
            if ($e->data['relation'] === 'member_of' && Sets::member($codes, $e->data['from'])) {
                $to = $e->data['to'];
                $k = self::pairKey($to);
                if (!isset($seen[$k])) {
                    $seen[$k] = true;
                    $out[] = $to;
                }
            }
        }
        usort($out, static fn (array $a, array $b): int => self::compareStringPair($a, $b));

        return $out;
    }

    /**
     * Substances via `contains` edges, grouped by the substance key that currently owns the `to`.
     *
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $edges
     * @param array<string, string> $subIndex
     * @return list<array<string,mixed>>
     */
    private static function resolveSubstances(array $codes, array $edges, array $subIndex): array
    {
        [$byKey, $keyOrder] = self::groupEdgesByOwner($edges, $codes, $subIndex, 'contains', 'from', 'to');

        $out = [];
        foreach ($byKey as $ok => $es) {
            $codesList = [];
            $codeSeen = [];
            foreach ($es as $e) {
                $ck = self::pairKey($e->data['to']);
                if (!isset($codeSeen[$ck])) {
                    $codeSeen[$ck] = true;
                    $codesList[] = $e->data['to'];
                }
            }
            usort($codesList, static fn (array $a, array $b): int => [$a[0], $a[1]] <=> [$b[0], $b[1]]);
            $out[] = ['key' => $keyOrder[$ok], 'codes' => $codesList, 'sources' => self::sortedUniqueSources($es)];
        }

        usort($out, static fn (array $a, array $b): int => self::compareOwner($a['key'], $b['key']));

        return $out;
    }

    /**
     * Depicted media via `depicts` edges (the first-class media-lane path, MED_* records).
     *
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $edges
     * @param array<string, array<string, array{0: string, 1: string}>> $mediaMembers
     * @param array<string, string> $mediaIndex
     * @param list<Claim> $attrs
     * @return list<array<string,mixed>>
     */
    private static function resolveDepicted(array $codes, array $edges, array $mediaMembers, array $mediaIndex, array $attrs, Priority|callable $priority): array
    {
        [$byKey, $keyOrder] = self::groupEdgesByOwner($edges, $codes, $mediaIndex, 'depicts', 'to', 'from');

        $out = [];
        foreach ($byKey as $ok => $es) {
            $owner = $keyOrder[$ok];
            $assetCodes = is_string($owner) && isset($mediaMembers[$owner])
                ? $mediaMembers[$owner]
                : Sets::of([self::ownerAsCode($owner)]);
            $attributes = self::laneAttributes($assetCodes, $attrs, $priority);

            $sources = self::sortedUniqueSources($es);

            $out[] = [
                'asset' => $owner,
                'role' => self::attrValue($attributes, 'role', 'secondary'),
                'source' => $sources[0],
                'uri' => self::attrValue($attributes, 'uri', null),
            ];
        }

        usort($out, static fn (array $a, array $b): int => self::compareOwner($a['asset'], $b['asset']));

        return $out;
    }

    /**
     * Derived descriptions (gr-sw0): descriptions tagged directly to the variant, plus
     * descriptions tagged to any substance it contains, minus steward-suppressed pairings.
     *
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $edges
     * @param array<string, array<string, array{0: string, 1: string}>> $lanes lane => members
     * @param array<string, string> $subIndex
     * @param array<string, string> $descIndex
     * @param list<array<string,mixed>> $substances this variant's resolveSubstances result
     * @param list<Claim> $attrs
     * @return list<array<string,mixed>>
     */
    private static function resolveDescriptions(array $codes, array $edges, array $lanes, array $subIndex, array $descIndex, array $substances, array $attrs, Priority|callable $priority): array
    {
        $describes = [];
        foreach ($edges as $e) {
            if ($e->data['relation'] === 'describes') {
                $describes[] = $e;
            }
        }

        // Substance keys this variant contains (the `via` source set).
        $contained = [];
        foreach ($substances as $sub) {
            $contained[self::ownerKey($sub['key'])] = true;
        }

        // direct: described codes the variant carries; via: described codes a contained substance owns.
        $entries = []; // list of [edge, route]
        foreach ($describes as $e) {
            if (Sets::member($codes, $e->data['to'])) {
                $entries[] = [$e, 'direct'];
            }
        }
        foreach ($describes as $e) {
            $owner = self::ownerOf($subIndex, $e->data['to']);
            $ok = self::ownerKey($owner);
            if (isset($contained[$ok])) {
                $entries[] = [$e, ['substance', $owner]];
            }
        }

        // Drop steward-suppressed pairings: a suppress edge (description code → carried product
        // code) hides every pairing resolving to that description record. One pass over the
        // edges builds the suppressed-key set; suppressed() rescanned them per entry.
        $suppressedKeys = [];
        foreach ($edges as $s) {
            if ($s->data['relation'] === 'suppress' && Sets::member($codes, $s->data['to'])) {
                $suppressedKeys[self::ownerKey(self::ownerOf($descIndex, $s->data['from']))] = true;
            }
        }
        $kept = [];
        foreach ($entries as $entry) {
            if (!isset($suppressedKeys[self::ownerKey(self::ownerOf($descIndex, $entry[0]->data['from']))])) {
                $kept[] = $entry;
            }
        }

        // Group by [owner-desc-key, route].
        /** @var array<string, array{key: mixed, route: mixed, entries: list<array{0: Claim, 1: mixed}>}> $groups */
        $groups = [];
        foreach ($kept as $entry) {
            [$e, $route] = $entry;
            $descOwner = self::ownerOf($descIndex, $e->data['from']);
            $gk = self::ownerKey($descOwner)."\x1e".self::routeKey($route);
            if (!isset($groups[$gk])) {
                $groups[$gk] = ['key' => $descOwner, 'route' => $route, 'entries' => []];
            }
            $groups[$gk]['entries'][] = $entry;
        }

        $out = [];
        foreach ($groups as $g) {
            $key = $g['key'];
            $descCodes = is_string($key) && isset($lanes['description'][$key])
                ? $lanes['description'][$key]
                : Sets::of([self::ownerAsCode($key)]);

            $out[] = [
                'key' => $key,
                'via' => $g['route'],
                'asserted_by' => self::sortedUniqueSources(array_column($g['entries'], 0)),
                'attributes' => self::laneAttributes($descCodes, $attrs, $priority),
            ];
        }

        // Sort by {via != :direct, via, key}.
        usort($out, static function (array $a, array $b): int {
            $da = $a['via'] === 'direct' ? 0 : 1;
            $db = $b['via'] === 'direct' ? 0 : 1;
            if ($da !== $db) {
                return $da <=> $db;
            }
            $cmp = strcmp(self::routeKey($a['via']), self::routeKey($b['via']));
            if ($cmp !== 0) {
                return $cmp;
            }

            return self::compareOwner($a['key'], $b['key']);
        });

        return $out;
    }

    /**
     * A lane's members flattened to code key => owning member key. First member wins, matching
     * the scan order the per-endpoint owner lookup used to walk.
     *
     * @param array<string, array<string, array{0: string, 1: string}>> $members
     * @return array<string, string>
     */
    private static function ownerIndex(array $members): array
    {
        $index = [];
        foreach ($members as $k => $set) {
            foreach ($set as $ck => $_code) {
                $index[$ck] ??= $k;
            }
        }

        return $index;
    }

    /**
     * Resolve an edge endpoint to the key that currently owns it; an endpoint with no identity
     * claim resolves to ITSELF (the code is the identity until a record exists).
     *
     * @param array<string, string> $index from {@see ownerIndex}
     * @param array{0: string, 1: string} $code
     * @return string|array{0: string, 1: string}
     */
    private static function ownerOf(array $index, array $code): string|array
    {
        return $index[Codes::key($code)] ?? $code;
    }

    /**
     * @param array<string, array{0: string, 1: string}> $codes
     * @param list<Claim> $attrs
     * @return list<array{0: string, 1: Decision}>
     */
    private static function laneAttributes(array $codes, array $attrs, Priority|callable $priority): array
    {
        $decisions = Survivorship::fieldDecisions($codes, $attrs, $priority);
        usort($decisions, static fn (array $a, array $b): int => strcmp($a[0], $b[0]));

        return $decisions;
    }

    /**
     * @param list<array{0: string, 1: Decision}> $attributes
     */
    private static function attrValue(array $attributes, string $field, mixed $default): mixed
    {
        foreach ($attributes as [$f, $decision]) {
            if ($f === $field) {
                return $decision->value;
            }
        }

        return $default;
    }

    // ── helpers ─────────────────────────────────────────────────────────────────

    /**
     * @param list<Claim> $claims
     * @return list<Claim>
     */
    private static function filterKind(array $claims, string $kind): array
    {
        $out = [];
        foreach ($claims as $c) {
            if ($c->kind === $kind) {
                $out[] = $c;
            }
        }

        return $out;
    }

    /** @param string|array{0: string, 1: string} $owner */
    private static function ownerKey(string|array $owner): string
    {
        return is_array($owner) ? self::pairKey($owner) : $owner;
    }

    /**
     * @param string|array{0: string, 1: string} $owner
     * @return array{0: string, 1: string}
     */
    private static function ownerAsCode(string|array $owner): array
    {
        // A description/media key resolving to itself: the owner key IS used as the lone "code".
        return is_array($owner) ? $owner : [$owner, $owner];
    }

    /**
     * @param string|array{0: string, 1: string} $a
     * @param string|array{0: string, 1: string} $b
     */
    private static function compareOwner(string|array $a, string|array $b): int
    {
        // Both substance/media/description owners are surrogate keys (strings) on every real path.
        return strcmp(self::ownerKey($a), self::ownerKey($b));
    }

    /** @param array{0: string, 1: string} $pair */
    private static function pairKey(array $pair): string
    {
        return $pair[0]."\x1f".$pair[1];
    }

    /**
     * Lexicographic comparison of two [scheme/collection, value] pairs — PHP's `<=>` would compare
     * numeric-looking strings NUMERICALLY ("1035" > "44"), but Elixir's term order is byte-wise.
     *
     * @param array{0: string, 1: string} $a
     * @param array{0: string, 1: string} $b
     */
    private static function compareStringPair(array $a, array $b): int
    {
        $c = strcmp($a[0], $b[0]);

        return $c !== 0 ? $c : strcmp($a[1], $b[1]);
    }

    private static function routeKey(mixed $route): string
    {
        if (is_string($route)) {
            return $route;
        }
        if (is_array($route)) {
            // ['substance', ownerKeyOrCode]
            return $route[0].':'.self::ownerKey($route[1]);
        }

        return (string) $route;
    }

    private static function scalarKey(mixed $value): string
    {
        if (is_array($value)) {
            // tuples like ['none','—'] or ['mpn','ALPHA']
            return 't:'.implode("\x1f", array_map(static fn ($x): string => (string) $x, $value));
        }

        return match (true) {
            is_bool($value) => 'b:'.($value ? '1' : '0'),
            is_int($value) => 'i:'.$value,
            is_float($value) => 'f:'.$value,
            is_string($value) => 's:'.$value,
            $value === null => 'n:',
            default => 'x:'.json_encode($value),
        };
    }

    private static function compareProductValues(mixed $a, mixed $b): int
    {
        // Elixir sorts product labels with the default term order. Real 422156 labels are integers
        // (the legacy entity); we compare numerically when both are ints, else by string form.
        if (is_int($a) && is_int($b)) {
            return $a <=> $b;
        }
        if (is_array($a) && is_array($b)) {
            return [$a[0] ?? '', $a[1] ?? ''] <=> [$b[0] ?? '', $b[1] ?? ''];
        }

        return strcmp(self::scalarKey($a), self::scalarKey($b));
    }

    /**
     * Group edges of $relation whose $memberEnd code the variant carries, keyed by the resolved
     * owner of their $ownerEnd — the shared shape of the substances and depicted resolvers.
     *
     * @param list<Claim> $edges
     * @param array<string, array{0: string, 1: string}> $codes
     * @param array<string, string> $index from {@see ownerIndex}
     * @return array{0: array<string, list<Claim>>, 1: array<string, string|array{0: string, 1: string}>}
     */
    private static function groupEdgesByOwner(array $edges, array $codes, array $index, string $relation, string $memberEnd, string $ownerEnd): array
    {
        $byKey = [];
        $keyOrder = [];
        foreach ($edges as $e) {
            if ($e->data['relation'] === $relation && Sets::member($codes, $e->data[$memberEnd])) {
                $owner = self::ownerOf($index, $e->data[$ownerEnd]);
                $ok = is_array($owner) ? self::pairKey($owner) : (string) $owner;
                if (!isset($byKey[$ok])) {
                    $byKey[$ok] = [];
                    $keyOrder[$ok] = $owner;
                }
                $byKey[$ok][] = $e;
            }
        }

        return [$byKey, $keyOrder];
    }

    /**
     * @param list<Claim> $claims
     * @return list<string> distinct sources, sorted
     */
    private static function sortedUniqueSources(array $claims): array
    {
        $out = [];
        $seen = [];
        foreach ($claims as $c) {
            if (!isset($seen[$c->source])) {
                $seen[$c->source] = true;
                $out[] = $c->source;
            }
        }
        sort($out, SORT_STRING);

        return $out;
    }
}

File: lib/golden_record/catalog.ex
defmodule GoldenRecord.Catalog do
  alias GoldenRecord.{Events, Lanes, Survivorship}

  # overrides: %{attr: %{{key, field} => ConflictResolved}, product: %{key => product}}
  #
  # `members` may span every lane (gr-2a8): the product lane becomes the variants below; the
  # other lanes feed the edge-traversal resolvers (substances/descriptions/depicted media) and
  # are projectable standalone via `lane_records/3`. Visibility is DERIVED at read time — a new
  # `contains` edge instantly pulls the substance's descriptions onto that product's page,
  # because nothing is copied; the projection is a fold (gr-sw0).
  def project(members, live_claims, priority, overrides) do
    lanes = Lanes.partition_members(members)
    attrs = Enum.filter(live_claims, &(&1.kind == :attribute))
    groups = Enum.filter(live_claims, &(&1.kind == :grouping))
    media = Enum.filter(live_claims, &(&1.kind == :media))
    edges = Enum.filter(live_claims, &(&1.kind == :edge))

    # Inverted code -> key maps, built once per projection so edge-endpoint resolution is a
    # lookup instead of a scan over the lane's members per edge.
    owners = %{
      substance: owner_index(lanes.substance),
      description: owner_index(lanes.description),
      media: owner_index(lanes.media)
    }

    # Suppress edges with their description endpoint pre-resolved; the product-code side is the
    # only part that varies per variant.
    suppress_edges =
      for s <- edges,
          s.data.relation == :suppress,
          do: {owner_key(owners.description, s.data.from), s.data.to}

    lanes.product
    |> Enum.map(fn {key, codes} ->
      substances = resolve_substances(codes, edges, owners.substance)

      %{
        key: key,
        codes: Enum.sort(MapSet.to_list(codes)),
        attributes: resolve_attributes(key, codes, attrs, priority, overrides.attr),
        product: resolve_product(key, codes, groups, priority, overrides.product),
        media:
          resolve_media(codes, media, priority) ++
            resolve_depicted(codes, edges, lanes.media, owners.media, attrs, priority),
        categories: resolve_categories(codes, edges),
        substances: substances,
        descriptions:
          resolve_descriptions(codes, lanes, owners, substances, suppress_edges, edges, attrs, priority)
      }
    end)
    |> Enum.group_by(& &1.product.value)
    |> Enum.sort_by(fn {product, _} -> product end)
    |> Enum.map(fn {product, vs} -> %{product: product, variants: Enum.sort_by(vs, & &1.key)} end)
  end

  @doc """
  Standalone view of a non-product lane's records (gr-2a8): each is a first-class golden record
  — identity codes, resolved attributes with survivorship — exactly what the product page embeds
  via edges, minus the traversal.
  """
  def lane_records(lane_members, live_claims, priority) do
    attrs = Enum.filter(live_claims, &(&1.kind == :attribute))

    lane_members
    |> Enum.map(fn {key, codes} ->
      %{
        key: key,
        codes: Enum.sort(MapSet.to_list(codes)),
        attributes: lane_attributes(codes, attrs, priority)
      }
    end)
    |> Enum.sort_by(& &1.key)
  end

  defp resolve_attributes(key, codes, attrs, priority, attr_overrides) do
    codes
    |> Survivorship.field_decisions(attrs, priority)
    |> Enum.map(fn {field, base} -> {field, apply_override(base, Map.get(attr_overrides, {key, field}))} end)
    |> Enum.sort()
  end

  defp apply_override(base, nil), do: base

  defp apply_override(base, %Events.ConflictResolved{decision: {:pick, v}, by: by}),
    do: %{base | value: v, winner: "steward:#{by}", status: :resolved_by_steward}

  defp resolve_product(key, codes, groups, priority, product_overrides) do
    case Map.get(product_overrides, key) do
      nil -> resolve_product_from_claims(codes, groups, priority)
      product -> %{value: product, winner: :steward, status: :resolved_by_steward, candidates: []}
    end
  end

  # Media attaches by code, exactly like an attribute — so it RE-HOMES automatically when a
  # split/merge moves its target code to a different surrogate key. Dedup by asset identity;
  # the highest-priority source wins each asset's metadata.
  defp resolve_media(codes, media, priority) do
    rank = Survivorship.rank_fun(priority)

    media
    |> Enum.filter(&MapSet.member?(codes, &1.data.target))
    |> Enum.group_by(& &1.data.asset)
    |> Enum.map(fn {asset, claims} ->
      best = Enum.min_by(claims, &rank.(:media, &1.source))
      %{asset: asset, role: best.data.role, source: best.source, uri: best.data.uri}
    end)
    |> Enum.sort_by(fn m -> {m.role != :primary, m.asset} end)
  end

  # Collection membership (e.g. ATC categories) is the :member_of edge relation (gr-xde): it
  # attaches by code so it re-homes on a split, and unions across sources by default.
  defp resolve_categories(codes, edges) do
    edges
    |> Enum.filter(&(&1.data.relation == :member_of and MapSet.member?(codes, &1.data.from)))
    |> Enum.map(& &1.data.to)
    |> Enum.uniq()
    |> Enum.sort()
  end

  # The substances this variant claims, via :contains edges — both hops resolve code → current
  # owner key at read time, so a substance merge converges every product's view with zero writes.
  defp resolve_substances(codes, edges, sub_index) do
    edges
    |> Enum.filter(&(&1.data.relation == :contains and MapSet.member?(codes, &1.data.from)))
    |> Enum.group_by(&owner_key(sub_index, &1.data.to))
    |> Enum.map(fn {key, es} ->
      %{
        key: key,
        codes: es |> Enum.map(& &1.data.to) |> Enum.uniq() |> Enum.sort(),
        sources: es |> Enum.map(& &1.source) |> Enum.uniq() |> Enum.sort()
      }
    end)
    |> Enum.sort_by(& &1.key)
  end

  # The derived description set (gr-sw0): descriptions tagged directly to this variant (one hop,
  # :describes), plus descriptions tagged to any substance it contains (two hops, via :contains)
  # — a named, bounded traversal, never blanket closure. Each entry carries its provenance (`via` — WHY it is
  # on this page) and drops steward-suppressed pairings (gr-745) for THIS product only.
  defp resolve_descriptions(codes, lanes, owners, substances, suppress_edges, edges, attrs, priority) do
    describes = Enum.filter(edges, &(&1.data.relation == :describes))
    contained = MapSet.new(substances, & &1.key)

    suppressed_keys =
      for {desc_key, to} <- suppress_edges, MapSet.member?(codes, to), into: MapSet.new(), do: desc_key

    direct = for e <- describes, MapSet.member?(codes, e.data.to), do: {e, :direct}

    via =
      for e <- describes,
          key = owner_key(owners.substance, e.data.to),
          MapSet.member?(contained, key),
          do: {e, {:substance, key}}

    (direct ++ via)
    |> Enum.reject(fn {e, _route} ->
      MapSet.member?(suppressed_keys, owner_key(owners.description, e.data.from))
    end)
    |> Enum.group_by(fn {e, route} -> {owner_key(owners.description, e.data.from), route} end)
    |> Enum.map(fn {{key, route}, entries} ->
      desc_codes = Map.get(lanes.description, key) || MapSet.new([key])

      %{
        key: key,
        via: route,
        asserted_by: entries |> Enum.map(fn {e, _} -> e.source end) |> Enum.uniq() |> Enum.sort(),
        attributes: lane_attributes(desc_codes, attrs, priority)
      }
    end)
    |> Enum.sort_by(&{&1.via != :direct, &1.via, &1.key})
  end

  # Media-lane records reach the page via :depicts edges — the first-class path (gr-kek). The
  # legacy :media claim kind keeps resolving in resolve_media/3 until every producer emits lanes.
  defp resolve_depicted(codes, edges, media_members, media_index, attrs, priority) do
    edges
    |> Enum.filter(&(&1.data.relation == :depicts and MapSet.member?(codes, &1.data.to)))
    |> Enum.group_by(&owner_key(media_index, &1.data.from))
    |> Enum.map(fn {key, es} ->
      attributes = lane_attributes(Map.get(media_members, key) || MapSet.new([key]), attrs, priority)

      %{
        asset: key,
        role: attr_value(attributes, "role", :secondary),
        source: es |> Enum.map(& &1.source) |> Enum.uniq() |> Enum.sort() |> hd(),
        uri: attr_value(attributes, "uri", nil)
      }
    end)
    |> Enum.sort_by(& &1.asset)
  end

  # Resolve an edge endpoint to the key that currently owns it; an endpoint with no identity
  # claim yet resolves to itself — the code IS the identity until a record exists for it.
  # put_new keeps the FIRST key that carries a code, matching the Enum.find_value scan the
  # index replaces (a held conflict can leave one code in two keys' sets).
  defp owner_index(members) do
    for {key, set} <- members, code <- set, reduce: %{} do
      acc -> Map.put_new(acc, code, key)
    end
  end

  defp owner_key(index, code), do: Map.get(index, code, code)

  defp lane_attributes(codes, attrs, priority),
    do: codes |> Survivorship.field_decisions(attrs, priority) |> Enum.sort()

  defp attr_value(attributes, field, default) do
    case List.keyfind(attributes, field, 0) do
      {_, %{value: v}} -> v
      nil -> default
    end
  end

  defp resolve_product_from_claims(codes, groups, priority) do
    groups
    |> Enum.filter(&MapSet.member?(codes, &1.data.code))
    |> Enum.map(&%{source: &1.source, value: &1.data.product, order: &1.order})
    |> case do
      [] ->
        %{value: {:none, "—"}, winner: nil, status: :resolved, candidates: []}

      entries ->
        base = Survivorship.decide(:product, entries, priority)
        # Contested: a single variant pointing at >1 product is a code collision — surface it.
        if entries |> Enum.map(& &1.value) |> Enum.uniq() |> length() > 1,
          do: %{base | value: nil, winner: nil, status: :needs_review},
          else: base
    end
  end
end
